src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -390,13 +413,109 @@ private static void CheckValueFormat(StringSegment value)
+
+            if (backSlashCount == 0)
+            {
+                return input;
+            }
+
+            var stringBuilder = new InplaceStringBuilder(input.Length - backSlashCount);
+
+            for (var i = 0; i < input.Length; i++)
+            {
+                if (input[i] == '\\')
+                {
+                    if (input.Length == i + 1)
+                    {
+                        throw new FormatException(string.Format(CultureInfo.InvariantCulture,
+                            "The header value is invalid. illegal escape character in input: '{0}', position: '{1}'", input, i + 1));
+                    }
+                    else
+                    {
+                        // Note, if a sender adds a quoted pair like '\\''n', we will assume it is over escaping and just add a n to the string.
+                        stringBuilder.Append(input[i + 1]);
+                        i++;
+                        continue;
+                    }
+                }
+                stringBuilder.Append(input[i]);
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        // See https://tools.ietf.org/html/rfc7230#section-3.2.6
+        private static StringSegment EscapeAsQuotedString(StringSegment input)
+        {
+            // By calling this, we know that the string requires quotes around it to be a valid token.
+            var backSlashCount = CountBackslashesForEncodingQuotedString(input);
+
+            var stringBuilder = new InplaceStringBuilder(input.Length + backSlashCount + 2); // 2 for quotes
+            stringBuilder.Append('\"');
+
+            for (var i = 0; i < input.Length; i++)
+            {
+                if (input[i] == '\\' || input[i] == '\"')
+                {
+                    stringBuilder.Append('\\');
+                }
+
+                stringBuilder.Append(input[i]);
+            }
+            stringBuilder.Append('\"');
+            return stringBuilder.ToString();
+        }
+
+        private static int CountBackslashesForDecodingQuotedString(StringSegment input)
+        {
+            var numBackSlashes = 0;
+            for (var i = 0; i < input.Length; i++)
+            {
+                if (input[i] == '\\')
+                {
+                    if (input.Length == i + 1)
+                    {
+                        throw new FormatException(string.Format(CultureInfo.InvariantCulture,
+                          "The header value is invalid. illegal escape character in input: '{0}', position: '{1}'", input, i + 1));
+                    }
+                    else if (input[i + 1] == '\\')
+                    {
+                        // Only count escaped backslashes once
+                        i++;
+                    }
+                    numBackSlashes++;
+                }
+            }
+            return numBackSlashes;
+        }
+
+        private static int CountBackslashesForEncodingQuotedString(StringSegment input)
```

Comment:
Name doesn't describe what the method does; actually counts backslashes and double-quotes. Suggest `CountCharactersNeedingBackslashesWhenEncoding()` or some such.